Fix support for bsc and heco

huazhouwang opened the pull request:
## What does this implement/fix? Explain your changes.
1. 加入 bsc 和 heco 的配置
2. HD 钱包支持 bsc 和 heco
3. 加入 token list


## Does this close any currently open issues?  
If it fixes a bug or resolves a feature request, be sure to link to that issue.
…

## Pull request type
<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 
_Put an `x` in the boxes that apply_
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## Where has this been tested?
…

## Any other comments?
…

shuaiLiWang commented:
这里边的第二个commit https://github.com/OneKeyHQ/electrum/pull/557的pr合进去就支持了

huazhouwang commented:
> 这里边的第二个commit #557 的pr合进去就支持了

#557 合进去还不能做到支持的

shuaiLiWang commented:
> > 这里边的第二个commit #557 的pr合进去就支持了
> 
> #557 合进去还不能做到支持的



> > 这里边的第二个commit #557 的pr合进去就支持了
> 
> #557 合进去还不能做到支持的

为什么呢 他那个pr都已经

> > 这里边的第二个commit #557 的pr合进去就支持了
> 
> #557 合进去还不能做到支持的

为什么呢 @taimanhui 

> > 这里边的第二个commit #557 的pr合进去就支持了
> 
> #557 合进去还不能做到支持的

@taimanhui  我记得是支持的 确认下

huazhouwang commented:
两个 commit 解决的问题是不同的，你是觉得哪些代码是重合的？

shuaiLiWang commented:
> 两个 commit 解决的问题是不同的，你是觉得哪些代码是重合的？@shuaiLiWang

第二个commit你解决的问题 都是合重的 
1.bsc和heco通过coin+address判断唯一
2.判断文件已存在时 #557 的修改也已经解决了这个问题

huazhouwang commented:
#557 那是写个新逻辑使用的，就逻辑没改。例如 get_unique_path 和 check_exist_file

shuaiLiWang commented:
> 是写给新逻辑使用的，旧逻辑没改

hunt的代码合进去以后 都是可以使用的代码 不存在新逻辑旧逻辑的 在557中check_exist_file已经不存在了 get_unique_path也已经挪到了helper中 @taimanhui 是吧

huazhouwang commented:
> hunt的代码合进去以后 都是可以使用的代码 不存在新逻辑旧逻辑的 在557中check_exist_file已经不存在了 get_unique_path也已经挪到了helper中 @taimanhui 是吧

#522 已经合并了，我 rebase 后看 check_exist_file 还在

taimanhui commented:
@huazhouwang `check_exist_file`在[这个commit](557/commits/680df974374e8779726bc15f0f915f4c07c24f75)里面去掉了，原因是每次循环判断一个wallet是否已经存在还是有点费事的。更新后的逻辑是这样的：

1. 从wallet的链+地址生成唯一的id，这一步在wallet实例化之后就可以得到，从而作为文件名并得到它对应的文件的路径
2. 已知id唯一，因此上一步得到的文件路径(path)唯一
3. daemon._wallets使用文件路径作为key（这一步和之前的逻辑仍然是一样的）
4. 判断那个path key是否在daemon._wallets里，或者使用`daemon.get_wallet(path)`得到的是否非None即可确认文件是否存在，从而不需要遍历那个daemon._wallets一个个现存的wallet比对地址和链

上面1和2是对后面要加的heco/bsc支持的漏洞修复，之前没考虑到不同链上相同地址，只使用地址生成那个id，添加bsc/heco几乎肯定会出问题
3和4是对原来逻辑的优化，其实原来把逻辑改成判断地址生成的那个hash是否在daemon._wallets里是一样的效果

huazhouwang commented:
OK， 我再 rebase 了，有两个地方需要改，就是之前提到的

taimanhui reviewed (commented):
- 第一个commit额外添加了个.DS_Store，如果你两个commit不合一起的话，rebase+edit处理一下它，另外两个commit的顺序应该是反过来的，先大致准备好再打开配置
- list_wallets的docstring需要更新体现增加heco
- 原来用来缓存钱包类型的命名的规则是"币种-XXXX"，但是原来只考虑btc和eth于是代码里有些对这个类型使用`[:3]`或者`[0:3]`然后进行比较的操作，console里一个，wallet_context里有两个
- wallet_context里面关于钱包列表的排序有个map需要增加heco

huazhouwang commented:
> * 原来用来缓存钱包类型的命名的规则是"币种-XXXX"，但是原来只考虑btc和eth于是代码里有些对这个类型使用`[:3]`或者`[0:3]`然后进行比较的操作，console里一个，wallet_context里有两个

那我改成 .split("-")[0]? 但我看有些地方是 .split("_")[0]，是分隔符改过？

shuaiLiWang commented:
> > * 原来用来缓存钱包类型的命名的规则是"币种-XXXX"，但是原来只考虑btc和eth于是代码里有些对这个类型使用`[:3]`或者`[0:3]`然后进行比较的操作，console里一个，wallet_context里有两个
> 
> 那我改成 .split("-")[0]? 但我看有些地方是 .split("_")[0]，是分隔符改过？

是 .split("-")[0]

huazhouwang commented:
目前看 wallet 级别和 global 级别用的分隔符不一样，所以打算分别用 `-` 和 `_` 拆解取出 coin

taimanhui commented:
> 目前看 wallet 级别和 global 级别用的分隔符不一样，所以打算分别用 `-` 和 `_` 拆解取出 coin

用config存wallet类型的这里面看起来用的都是`-`，你说的用下划线的地方应该不在这个范围里，还是有另外的地方会用下划线然后也写进了这个里面，那就得都考虑了

huazhouwang commented on electrum_gui/android/console.py:
```diff
@@ -528,7 +528,7 @@ def load_wallet(self, name, password=None):
                 return
             wallet_type = db.data["wallet_type"]
-            if wallet_type[0:3] in self.coins:
+            if wallet_context.split_coin_from_wallet_type(wallet_type) in self.coins:
```
@taimanhui 这里用的是下划线

huazhouwang commented:
我加了个新的 commit，加入 token list

taimanhui commented on electrum_gui/android/wallet_context.py:
```diff
@@ -11,8 +12,15 @@
     "btc": 3,
     "eth": 2,
     "bsc": 1,
+    "heco": 1,
```
这样的话bsc和heco会混排的哦，确定可以？

因为排序的时候还考虑时间戳所以这里只能是大的数表示在前，估计我们得往下列到负数…

huazhouwang replied:
目前需求没说这俩有先后关系，所以就按时间顺序来了

taimanhui commented on electrum_gui/android/wallet_context.py:
```diff
@@ -11,8 +12,15 @@
+    "heco": 1,
 }
 
+_WALLET_TYPE_DIVIDE_PATTERN = re.compile(r"[_\-]")
```
我记得`-`在最前和最后的话不用转义的，不过我测试了下也不会误认那个`\`，估计还是要两个才有效果

huazhouwang replied:
嗯，测试是没问题的

taimanhui reviewed (approved):
拍戏那里需要看是否允许heco和bsc混排，不然的话他们的顺序是怎么样的

huazhouwang commented:
@taimanhui 第四个 commit 帮忙看下，BSC 和 Heco 的 ERC20 加载问题

taimanhui commented on electrum/eth_wallet.py:
```diff
@@ -1183,8 +1185,11 @@ class Standard_Eth_Wallet(Simple_Eth_Deterministic_Wallet):
+
         wallet = cls(db, None, config=config, index=index)
         wallet.coin = coin
         wallet.wallet_type = f"{coin}_standard"
```
这行不需要了，但是好像去掉又很不明确，注释掉它然后说明已经通过db读取到了类型了？或者在上面`put wallet_type`的地方说明一下比较好。

另外`Imported_Eth_Wallet`里也有两处类似的，看下是不是也需要改。

直接用类变量区分币种看起来在这种共用的情况下比较混乱，我拿个小本本记下看后面怎么改...

huazhouwang replied:
其实这些属性跟钱包无关。
如果说后续想要拿 DB 来管理的话，应该是：

钱包：
  - 币种
  - 地址

地址：
  - 类型
  - 币种
  - address str
  - pubkey id(如果有)
  - bip44 path（如果有）

pubkey:
  - 类型(单公钥，从 private key 来，xpub)
  - key str
  - secret key id

secret key：
  - 类型（单私钥，xprv, seed）
  - key str

简单描述就这些，通过这样定义结构，可以便捷地支持各种导入类型。可以随时讨论这块

taimanhui replied:
> 其实这些属性跟钱包无关。

也不是完全无关，而且像这种用来内部管理的类型在使用这个底层文件做配置的机制被改动之前都还得用这个wallet db，原因是用其他的再来处理它们就有多个地方要协作处理一件事情（比如从底层文件发现并加载）从而涉及一致性的处理。

另外目前看底层这个存储钱包的机制就算要改成用数据库的方式也肯定放到更后面比较合适。

> 如果说后续想要拿 DB 来管理的话，应该是：
> 钱包：
>   - 币种

我理解这里的币种实际上应该指链？或者是狭义上某条链的基础币种（比如eth之于ethereum，我不晓得有没专门名词表达这个意思）

另外我上面的comment说有其他几处生成钱包的classmethod也要先把wallet type设置到db里面你的回复没涉及这个，提醒下别漏了看是不是需要哦

huazhouwang replied:
> 我理解这里的币种实际上应该指链？

指 chain_code

> 我上面的comment说有其他几处生成钱包的classmethod也要先把wallet type设置到db里面你的回复没涉及这个，提醒下别漏了看是不是需要哦

嗯，在看

huazhouwang commented:
@taimanhui 第5个 commit

taimanhui commented on electrum/eth_wallet.py:
```diff
@@ -1272,7 +1277,15 @@ def register_wallet_type(category):
+
+    'heco_standard': Standard_Eth_Wallet,
+    'heco_xpub': Standard_Eth_Wallet,
+    'heco_imported': Imported_Eth_Wallet
```
既然这里改到了如果codestyle允许的话，补个`,`在最后吧

huazhouwang replied:
ack
